identity_document/src/document/core_document.rs

```diff
@@ -865,12 +807,67 @@ where
       None => self.data.verification_method.query_mut(query),
     }
   }
+
+  /// Update the DID components of the document's `id`, controllers, methods and services by applying the provided
+  /// fallible maps.
+  ///
+  /// This is an advanced method that can be useful for DID methods that do not know the document's identifier prior
+  /// to publishing, but should preferably be avoided otherwise.
+  ///
+  /// # Errors
+  /// Any error is returned if any of the functions fail or the updates cause scoped method references to embedded
+  /// methods, or methods and services with identical identifiers in the document. In the case where illegal identifiers
+  /// are detected the supplied the `error_cast` function gets called in order to convert [`Error`] to `E`.
+  pub fn try_update_identifiers<F, G, H, L, M, E>(
+    self,
+    id_update: F,
+    controller_update: G,
+    methods_update: H,
+    service_update: L,
+    error_cast: M,
+  ) -> Result<Self, E>
+  where
+    F: FnOnce(CoreDID) -> std::result::Result<CoreDID, E>,
+    G: FnMut(CoreDID) -> std::result::Result<CoreDID, E>,
+    H: FnMut(CoreDID) -> std::result::Result<CoreDID, E>,
+    L: FnMut(CoreDID) -> std::result::Result<CoreDID, E>,
+    M: FnOnce(crate::Error) -> E,
+  {
+    let data = self
+      .data
+      .try_update_identifiers(id_update, controller_update, methods_update, service_update)?;
+    CoreDocument::try_from(data).map_err(error_cast)
+  }
```

Comment:
Do we absolutely need this method?

Reply:
I tried to remove all of these maps, but since we need to mutate identifiers when resolving an `IotaDocument` we need some way of doing that. I would rather  do it to this than add methods to `OrderedSet` that make it possible to break its invariants.

Reply:
Ah, sorry the changed files made it look like there were just two versions of this function, but one is actually a private fn on `CoreDocumentData`.
So I'm fine with exposing this in general, but the main point here was whether the error_cast is really necessary? I'd be fine with just requiring `Result<CoreDID, identity_document::Error>` to not add even more generics on that method.